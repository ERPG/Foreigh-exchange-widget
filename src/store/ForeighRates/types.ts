import { ForeighRates } from '../../models/ForeighRates';

export const GET_FOREIGH_RATES = 'GET_FOREIGH_RATES';
export const GENERETATE_FOREIGH_RATES_ERROR = 'GENERETATE_FOREIGH_RATES_ERROR';

export interface GetForeighRatesAction {
  type: typeof GET_FOREIGH_RATES;
  payload: ForeighRates;
}

export interface GenerateForeighRatesErrorAction {
  type: typeof GENERETATE_FOREIGH_RATES_ERROR;
  error: Error;
}

export type ForeighRatesActionTypes =
  | GetForeighRatesAction
  | GenerateForeighRatesErrorAction;
