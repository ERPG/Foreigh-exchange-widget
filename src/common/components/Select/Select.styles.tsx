import { makeStyles, Theme } from '@material-ui/core';

export const useSelectStyles = makeStyles((theme: Theme) => ({
  root: {
    ...theme.typography.body1,
  },
}));
