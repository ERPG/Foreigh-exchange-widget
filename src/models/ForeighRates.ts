export enum Currencies {
  EUR = 'EUR',
  USD = 'USD',
  GBP = 'GBP',
}

export type Rates<R> = { [key in keyof typeof Currencies]: R };

export type CurrencyRate<T> = { currency: Currencies; rate: T };

export interface IForeighRatesPayload {
  base: Currencies;
  rates: Rates<string>;
}

export class ForeighRates {
  private readonly _base: Currencies;
  private readonly _rates: Rates<string>;

  constructor(foreighRates: IForeighRatesPayload) {
    this._rates = foreighRates.rates;
    this._base = foreighRates.base;
  }

  get base(): Currencies {
    return this._base;
  }

  get rates(): Rates<string> {
    return this._rates;
  }

  public getExchangeRatesArr() {
    let newRatesArr: CurrencyRate<number>[] = [];
    for (const key in this._rates) {
      newRatesArr.push({
        currency: key as Currencies,
        // == CAMBIAR EL ANY ==
        rate: (this._rates as any)[key],
      });
    }
    return newRatesArr;
  }
}
