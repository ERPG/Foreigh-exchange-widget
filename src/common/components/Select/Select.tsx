import {
  Select as MuiSelect,
  SelectChangeEvent,
  MenuItem,
  FormControl,
} from '@mui/material';
import { CurrencyRate } from '../../../models/ForeighRates';
import { useSelectStyles } from './Select.styles';

export interface ISelectProps {
  value: number;
  onChange: (e: SelectChangeEvent<string>) => void;
  items: CurrencyRate<number>[] | undefined;
  name: string;
}

const Select = ({ value, onChange, items, name }: ISelectProps) => {
  const classes = useSelectStyles();

  return (
    <FormControl variant="standard" sx={{ m: 1, minWidth: 120 }} fullWidth>
      <MuiSelect
        classes={{ root: classes.root }}
        name={name}
        onChange={onChange}
        inputProps={{
          'aria-label': 'Without label',
          'data-fx': 'fx-currency-selection',
        }}
        value={value.toString()}
        defaultValue={value.toString()}
      >
        {items?.map((elem: CurrencyRate<number>, index: number) => (
          <MenuItem key={index} value={elem.rate.toString()}>
            {elem.currency}
          </MenuItem>
        ))}
      </MuiSelect>
    </FormControl>
  );
};

export default Select;
