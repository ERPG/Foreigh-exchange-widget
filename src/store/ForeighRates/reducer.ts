import { AppState } from '..';
import { ForeighRates } from '../../models/ForeighRates';
import {
  ForeighRatesActionTypes,
  GENERETATE_FOREIGH_RATES_ERROR,
  GET_FOREIGH_RATES,
} from './types';

export interface IForeighRatesState {
  item?: ForeighRates;
  error?: Error;
}

export const INITIAL_STATE: IForeighRatesState = { item: undefined };

export function ForeighRatesReducer(
  state = INITIAL_STATE,
  action: ForeighRatesActionTypes
): IForeighRatesState {
  switch (action.type) {
    case GET_FOREIGH_RATES:
      return { ...state, item: action.payload };
    case GENERETATE_FOREIGH_RATES_ERROR:
      return { ...state, error: action.error };
    default:
      return state;
  }
}

export const selectForeighRates = (state: AppState): ForeighRates | undefined =>
  state.item;
