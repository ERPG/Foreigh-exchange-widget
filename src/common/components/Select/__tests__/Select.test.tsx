import { render, fireEvent, cleanup } from '@testing-library/react';
import { Currencies } from '../../../../models/ForeighRates';
import { customRender } from '../../../../../__tests__/__utils__/testUtils';
import Select from '../Select';

describe('Select component', () => {
  const mockOnchange = jest.fn();
  const defaultProps = {
    name: 'test-select',
    value: 0,
    onChange: mockOnchange,
    items: [
      { currency: Currencies.EUR, rate: 1 },
      { currency: Currencies.USD, rate: 2 },
    ],
  };

  afterEach(cleanup);

  it('Should render component without errors', () => {
    const { container } = render(<Select {...defaultProps} />);

    expect(container).toMatchSnapshot();
  });

  it('Should display and change options', () => {
    const { getByDataFx } = customRender(<Select {...defaultProps} />);

    const selectEl = getByDataFx('fx-currency-selection');
    expect(selectEl).toBeDefined();

    fireEvent.change(selectEl, { target: { value: '1' } });

    expect(mockOnchange).toHaveBeenCalled();
  });
});
