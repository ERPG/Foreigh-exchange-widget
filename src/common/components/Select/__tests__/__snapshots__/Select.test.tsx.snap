// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select component Should render component without errors 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth css-1v3hyuc-MuiFormControl-root"
  >
    <div
      class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl makeStyles-root-1 MuiSelect-root css-1480iag-MuiInputBase-root-MuiInput-root"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-label="Without label"
        aria-labelledby="mui-component-select-test-select"
        class="MuiSelect-select MuiSelect-standard MuiInput-input MuiInputBase-input css-1rxz5jq-MuiSelect-select-MuiInputBase-input-MuiInput-input"
        id="mui-component-select-test-select"
        role="button"
        tabindex="0"
      >
        <span
          class="notranslate"
        >
          â€‹
        </span>
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        data-fx="fx-currency-selection"
        name="test-select"
        tabindex="-1"
        value="0"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
</div>
`;
